File: windows-launcher/renderer/artifact-detector.js
// Pure, DOM-free logic for deciding whether a chat reply contains content
// substantial enough to deserve its own standalone view instead of being
// crammed into a chat bubble -- same testable-module split as
// avatar/live2d-logic.js. The actual rendering (markdown-render.js) needs
// a real DOM (DOMPurify), so it isn't unit-tested the same way; this part
// -- the "should we" decision -- doesn't need one and is fully covered.

// An explicit ```html or ```mermaid fence is always artifact-worthy
// regardless of size (both are meant to be viewed/rendered, not read as
// chat text -- a compact 5-node flowchart is often well under the length
// threshold below); any other fenced block only qualifies once it's long
// enough that inlining it would dominate the chat bubble.
const ARTIFACT_MIN_CHARS = 400;
const ALWAYS_ARTIFACT_LANGUAGES = new Set(["html", "mermaid"]);

// Returns the first artifact-worthy fenced block in `markdownText`, or
// null. Only the first match -- a reply naming a second one is treated as
// chat content, not a second artifact.
function extractArtifact(markdownText) {
  const text = String(markdownText || "");
  const fenceRegex = /```(\w*)\r?\n([\s\S]*?)```/g;
  let match;
  while ((match = fenceRegex.exec(text))) {
    const language = (match[1] || "").toLowerCase();
    const content = match[2];
    if (ALWAYS_ARTIFACT_LANGUAGES.has(language) || content.length >= ARTIFACT_MIN_CHARS) {
      return {
        language: language || "text",
        content: content.replace(/\s+$/, ""),
        matchedText: match[0],
      };
    }
  }
  return null;
}

// Issue #391: groups a newly detected artifact into the version "thread" of
// the most recent artifact of the same language in `history` (the ordered
// list of artifacts already assigned this session), when the two share
// enough content to plausibly be revisions of one thing rather than two
// unrelated artifacts that happen to use the same language.
// ponytail: a fixed line-overlap threshold, not real content-identity --
// there's no stable id the model provides for "this is the same artifact as
// before." Upgrade path is a model-supplied artifact id/title if this
// heuristic ever misfires often enough in practice to matter.
const SAME_ARTIFACT_LINE_OVERLAP_THRESHOLD = 0.3;

function lineOverlapRatio(contentA, contentB) {
  const linesA = new Set(contentA.split("\n").map((line) => line.trim()).filter(Boolean));
  const linesB = new Set(contentB.split("\n").map((line) => line.trim()).filter(Boolean));
  if (!linesA.size || !linesB.size) return 0;
  let shared = 0;
  for (const line of linesA) {
    if (linesB.has(line)) shared += 1;
  }
  return shared / Math.max(linesA.size, linesB.size);
}

// Returns `artifact` enriched with `threadId` (which version-thread it
// belongs to) and `versionIndex` (1-based position within that thread).
// Reuses the most recent same-language entry in `history`'s threadId when
// the content overlaps enough to look like a revision; otherwise starts a
// new thread. `history` is not mutated -- the caller decides whether/where
// to store the returned, enriched artifact.
function assignArtifactVersion(artifact, history) {
  let lastSameLanguage = null;
  for (let i = history.length - 1; i >= 0; i--) {
    if (history[i].language === artifact.language) {
      lastSameLanguage = history[i];
      break;
    }
  }
  const isNewVersion =
    lastSameLanguage &&
    lineOverlapRatio(artifact.content, lastSameLanguage.content) >= SAME_ARTIFACT_LINE_OVERLAP_THRESHOLD;
  const threadId = isNewVersion ? lastSameLanguage.threadId : `${artifact.language}-${history.length}`;
  const versionIndex = isNewVersion
    ? history.filter((a) => a.threadId === threadId).length + 1
    : 1;
  return { ...artifact, threadId, versionIndex };
}

module.exports = {
  ARTIFACT_MIN_CHARS,
  ALWAYS_ARTIFACT_LANGUAGES,
  extractArtifact,
  assignArtifactVersion,
  SAME_ARTIFACT_LINE_OVERLAP_THRESHOLD,
};

File: windows-native-launcher/ArtifactDetector.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace Mana.NativeLauncher;

internal readonly record struct DetectedArtifact(string Language, string Content, string MatchedText);

internal readonly record struct VersionedArtifact(string Language, string Content, string ThreadId, int VersionIndex);

// #528: ports windows-launcher/renderer/artifact-detector.js verbatim --
// pure, DOM-free logic deciding whether a reply contains content
// substantial enough for its own standalone viewer, and which "version
// thread" a newly detected artifact belongs to. The actual rendering
// (MermaidParser/MermaidLayout/ArtifactViewerForm) is untested WinForms/
// GDI+-coupled code, same split the reference itself uses (this part
// doesn't need a DOM, the renderer does).
internal static class ArtifactDetector
{
    public const int ArtifactMinChars = 400;
    private static readonly HashSet<string> AlwaysArtifactLanguages = new(System.StringComparer.Ordinal) { "html", "mermaid" };
    private const double SameArtifactLineOverlapThreshold = 0.3;

    private static readonly Regex FencePattern = new(@"```(\w*)\r?\n([\s\S]*?)```", RegexOptions.Compiled);

    // Returns the first artifact-worthy fenced block in text, or null --
    // only the first match, a reply naming a second one is treated as
    // chat content, not a second artifact.
    public static DetectedArtifact? Extract(string? markdownText)
    {
        var text = markdownText ?? "";
        foreach (Match match in FencePattern.Matches(text))
        {
            var language = match.Groups[1].Value.ToLowerInvariant();
            var content = match.Groups[2].Value;
            if (AlwaysArtifactLanguages.Contains(language) || content.Length >= ArtifactMinChars)
            {
                return new DetectedArtifact(
                    language.Length > 0 ? language : "text",
                    content.TrimEnd(),
                    match.Value);
            }
        }
        return null;
    }

    // Groups a newly detected artifact into the version thread of the
    // most recent same-language entry in history, when the two share
    // enough content to plausibly be revisions of one thing. history is
    // not mutated -- the caller decides whether/where to store the
    // returned, enriched artifact.
    //
    // A module-level, ever-incrementing thread counter (not derived from
    // history.Count) -- matches the reference's own reasoning: a caller
    // may thread against more than one independent history list, and a
    // counter tied to any single list's length could produce the same
    // threadId for two genuinely unrelated artifacts once merged.
    private static int nextNewThreadId;

    public static VersionedArtifact AssignVersion(DetectedArtifact artifact, IReadOnlyList<VersionedArtifact> history)
    {
        VersionedArtifact? lastSameLanguage = null;
        for (var i = history.Count - 1; i >= 0; i--)
        {
            if (history[i].Language == artifact.Language)
            {
                lastSameLanguage = history[i];
                break;
            }
        }

        var isNewVersion = lastSameLanguage is { } last
            && LineOverlapRatio(artifact.Content, last.Content) >= SameArtifactLineOverlapThreshold;
        var threadId = isNewVersion ? lastSameLanguage!.Value.ThreadId : $"{artifact.Language}-{nextNewThreadId++}";

        var versionIndex = 1;
        if (isNewVersion)
        {
            var count = 0;
            foreach (var entry in history)
            {
                if (entry.ThreadId == threadId)
                {
                    count++;
                }
            }
            versionIndex = count + 1;
        }

        return new VersionedArtifact(artifact.Language, artifact.Content, threadId, versionIndex);
    }

    private static double LineOverlapRatio(string contentA, string contentB)
    {
        var linesA = NonEmptyTrimmedLines(contentA);
        var linesB = NonEmptyTrimmedLines(contentB);
        if (linesA.Count == 0 || linesB.Count == 0)
        {
            return 0;
        }

        var shared = 0;
        foreach (var line in linesA)
        {
            if (linesB.Contains(line))
            {
                shared++;
            }
        }
        return (double)shared / System.Math.Max(linesA.Count, linesB.Count);
    }

    private static HashSet<string> NonEmptyTrimmedLines(string content)
    {
        var lines = new HashSet<string>();
        foreach (var line in content.Split('\n'))
        {
            var trimmed = line.Trim();
            if (trimmed.Length > 0)
            {
                lines.Add(trimmed);
            }
        }
        return lines;
    }
}
